src/lib/storage/reference_segment.cpp

```diff
@@ -6,38 +6,44 @@
 namespace opossum {
 
 ReferenceSegment::ReferenceSegment(const std::shared_ptr<const Table>& referenced_table,
-                                   const ColumnID referenced_column_id, const std::shared_ptr<const PosList>& pos) {
-  // Implementation goes here
-}
+                                   const ColumnID referenced_column_id, const std::shared_ptr<const PosList>& pos)
+    : _referenced_table{referenced_table}, _pos{pos}, _referenced_column_id{referenced_column_id} {}
 
 AllTypeVariant ReferenceSegment::operator[](const ChunkOffset chunk_offset) const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  Assert(chunk_offset < size(), "Chunk offset " + std::to_string(chunk_offset) + " is out of bounds.");
+  const auto row_id = pos_list()->operator[](chunk_offset);
+
+  if (row_id.is_null()) {
+    return NULL_VALUE;
+  }
+
+  const auto chunk = referenced_table()->get_chunk(row_id.chunk_id);
+  const auto segment = chunk->get_segment(referenced_column_id());
+
+  DebugAssert(!std::dynamic_pointer_cast<ReferenceSegment>(segment),
+              "ReferenceSegment only supports referencing ValueSegment or DictionarySegment.");
+
+  return segment->operator[](row_id.chunk_offset);
 }
 
 ChunkOffset ReferenceSegment::size() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return pos_list()->size();
 }
 
 const std::shared_ptr<const PosList>& ReferenceSegment::pos_list() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _pos;
 }
 
 const std::shared_ptr<const Table>& ReferenceSegment::referenced_table() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _referenced_table;
 }
 
 ColumnID ReferenceSegment::referenced_column_id() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return _referenced_column_id;
 }
 
 size_t ReferenceSegment::estimate_memory_usage() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return sizeof(*this);
```

Comment:
Interesting approach. This would not count the concrete size of the referenced PosList, right? If we wanted to profile the whole memory usage of the db, where would you count that memory?